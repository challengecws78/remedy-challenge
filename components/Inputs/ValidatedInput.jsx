import React, { Component } from 'react';


export class ValidatedInput extends Component {
    constructor(props) {
        super(props)
        this.state = {
            value: props.value,
            error: false
        }
    }

    static getDerivedStateFromProps(nextProps) {
        return { value: nextProps.value }
    }

    handleChange = (e) => {
        const name = this.props.name;
        const { value } = e.target
        const error = this.props.validate ? this.props.validate(value) : false;
        this.setState({ value, error });
        this.props.onChange({ name, value });
    }

    get returnInput () {
        return (
            <div className="form-group">
                <label>{this.props.label}</label>
                <input
                    name={this.props.name}
                    type={this.props.type}
                    className="form-control"
                    placeholder={this.props.placeholder}
                    value={this.state.value}
                    onChange={this.handleChange}                   
                />
                <span style={{ color: 'red' }}>{this.state.error}</span>            
            </div>
        )
    }

    get returnTextArea() {
        return (
            <div className="form-group">
                <label>{this.props.label}</label>
                <textarea
                    name={this.props.name}
                    type={this.props.type}
                    className="form-control"
                    placeholder={this.props.placeholder}
                    value={this.state.value}
                    onChange={this.handleChange}                   
                />
                <span style={{ color: 'red' }}>{this.state.error}</span>            
            </div>
        )
    }

    render() {
        return (
                this.props.textArea ? this.returnTextArea : this.returnInput
        )
    }
}

export default ValidatedInput;