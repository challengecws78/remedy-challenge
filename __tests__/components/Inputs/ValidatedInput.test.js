import React from 'react';
import { ValidatedInput } from '../../../components/Inputs/ValidatedInput';
import { shallow } from 'enzyme';

describe('<ValidatedInput />', () => {
    it('should render ValidatedInput as an input', () => {
        const onChange = jest.fn();
        const validate = jest.fn();
        const wrapper = shallow(
            <ValidatedInput 
                label="Name" 
                name="name"
                type="text"
                placeholder="Name"
                value=''
                textArea={false}
                validate={validate}
                onChange={onChange}
            />
        ); 
        expect(wrapper).toMatchSnapshot();
    })

    it('should render ValidatedInput as a textarea', () => {
        const onChange = jest.fn();
        const validate = jest.fn();
        const wrapper = shallow(
            <ValidatedInput 
                label="Message" 
                name="message"
                type="text"
                placeholder="Message"
                value=''
                textArea={true}
                validate={validate}
                onChange={onChange}
            />
        ); 
        expect(wrapper).toMatchSnapshot();
    })

    it('should throw error if name is empty', () => {
        const onChange = jest.fn();
        const name = 'name'
        const value = '';
        const wrapper = shallow(
            <ValidatedInput
                label="Name" 
                name={name}
                type="text"
                placeholder="Name"
                value={value}
                validate={val => (val ? false : 'Name is Required')}
                onChange={onChange}
            />
        ); 
        wrapper.find('input').simulate('change', {
        target: { name, value }
        });
        expect(wrapper.state('error')).toBe('Name is Required');
    })

    it('should not error if name is not empty', ()=> {
        const onChange = jest.fn();
        const name = 'name'
        const value = 'Aaron Rodgers';
        const wrapper = shallow(
            <ValidatedInput
                label="Name" 
                name={name}
                type="text"
                placeholder="First Name"
                value={value}
                validate={val => (val ? false : 'Name is Required')}
                onChange={onChange}
            />
        ); 
        wrapper.find('input').simulate('change', {
          target: { name, value }
        });
        expect(wrapper.state('error')).toBe(false);
    });
});