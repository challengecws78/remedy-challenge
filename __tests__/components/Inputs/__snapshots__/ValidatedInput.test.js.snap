// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidatedInput /> should render ValidatedInput as a textarea 1`] = `
<div
  className="form-group"
>
  <label>
    Message
  </label>
  <textarea
    className="form-control"
    name="message"
    onChange={[Function]}
    placeholder="Message"
    type="text"
    value=""
  />
  <span
    style={
      Object {
        "color": "red",
      }
    }
  />
</div>
`;

exports[`<ValidatedInput /> should render ValidatedInput as an input 1`] = `
<div
  className="form-group"
>
  <label>
    Name
  </label>
  <input
    className="form-control"
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <span
    style={
      Object {
        "color": "red",
      }
    }
  />
</div>
`;
